// @alexa-games/sfb-f/test/utilities/storyUtilities.ts
import * as fs from 'fs';
import * as path from 'path';

export const POSITIVE_TEST_STORY_DIRECTORY = './test/positive_test_cases';
export const NEGATIVE_TEST_STORY_DIRECTORY_MISSING_START = './test/negative_test_cases/missing_start_scene';
export const GLOBAL_TEST_STORY_DIRECTORY = './test/global_test_cases';

export interface Content {
  id: string,
  text: string
}

export function loadTestStory(storyFileDirectory: string, fileName: string): string {
    const fullPath = path.resolve(path.join(storyFileDirectory, fileName));

    // console.log(`fullPath=${fullPath}`);

    return fs.readFileSync(fullPath, {
        encoding: "utf8"
      });
}

export function loadTestResource(storyFileDirectory: string, fileName: string): string {
    const fullPath = path.resolve(path.join(storyFileDirectory, 'resources', fileName));

    // console.log(`fullPath=${fullPath}`);

    return fs.readFileSync(fullPath, {
        encoding: "utf8"
      });
}

export function loadTestResourceAsObject(storyFileDirectory: string, fileName: string): any {
    const fullPath = path.resolve(path.join(storyFileDirectory, 'resources', fileName));

    // console.log(`fullPath=${fullPath}`);

    return JSON.parse(fs.readFileSync(fullPath, {
        encoding: "utf8"
      }));
}

export function loadAllContent(storyFileDirectory: string, storyFileList: string[]): Content[] {
    const contentList: Content[] = [];

    for (let storyFile of storyFileList) {
      contentList.push({
        id: storyFile,
        text: loadTestStory(storyFileDirectory, storyFile)
      });
    }

    return contentList;
}


export function getISPMapping(storyFileDirectory: string, ispFile: string): { [key: string]: string } {
    let ispJSON = JSON.parse(loadTestResource(storyFileDirectory, ispFile));

    let ispMapping: { [key: string]: string } = {};
    for (let i: number = 0; i < ispJSON.length; i++) {
      ispMapping[ispJSON[i].productName] = ispJSON[i].ISPID;
    }
    return ispMapping;
}
