// @alexa-games/sfb-skill/src/aceExtension/alexaAPLExtension.ts
import { DriverExtension, DriverExtensionParameter, VisualOptions, ImportErrorLine } from '@alexa-games/sfb-f';
import { ConfigAccessor } from './../configAccessor';
import { APLHelper } from './../handler/APLHelper';

export class AlexaAPLExtension implements DriverExtension {
    private aplHelper: APLHelper;
    constructor(locale: string, configAccessor: ConfigAccessor) {
        try {
            this.aplHelper = new APLHelper(locale, configAccessor);
        } catch (err) {
            throw <ImportErrorLine>{
                errorMessage: err,
                errorName: 'APL Extension Error',
                lineNumber: 0,
                sourceID: "APL_TEMPLATE"
            };
        }
    }

    async post(param: DriverExtensionParameter) {
        const handlerInput = param.userInputHelper.getHandlerInput();
        const driver = param.driver;

        let visualProperties: VisualOptions[] | undefined = await driver.getVisuals();

		if (handlerInput && visualProperties && this.aplHelper.supportsDisplay(handlerInput)) {
			let visualDirectives: any[] = this.aplHelper.generateAPLDirectiveWithVisualOptions(visualProperties);
			for (let aplDirective of visualDirectives) {
				handlerInput.responseBuilder.addDirective(aplDirective);
			}
		}
    }

    async pre(param: DriverExtensionParameter) {
    }
}
