// @alexa-games/sfb-f/dist/aceEntities/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuiltInUserInput = {
    "Resume Story": {
        intent: "LaunchRequest",
    },
    "Exit Story": {
        intent: "ExitRequest"
    }
};
var BuiltInScenes;
(function (BuiltInScenes) {
    BuiltInScenes["ResumeScene"] = "Resume";
    BuiltInScenes["PauseScene"] = "Pause";
    BuiltInScenes["StartScene"] = "Start";
})(BuiltInScenes = exports.BuiltInScenes || (exports.BuiltInScenes = {}));
var AudioBlendOption;
(function (AudioBlendOption) {
    AudioBlendOption["TRIM"] = "trim";
    AudioBlendOption["LONGEST"] = "longest";
})(AudioBlendOption = exports.AudioBlendOption || (exports.AudioBlendOption = {}));
var AlexaAudioType;
(function (AlexaAudioType) {
    AlexaAudioType["POLLY"] = "polly";
    AlexaAudioType["AUDIO"] = "audio";
    AlexaAudioType["TEXT"] = "text";
})(AlexaAudioType = exports.AlexaAudioType || (exports.AlexaAudioType = {}));
var SceneDirectionType;
(function (SceneDirectionType) {
    SceneDirectionType["CONDITION"] = "condition";
    SceneDirectionType["CHOICE"] = "choice";
    SceneDirectionType["BGM"] = "backgroundmusic";
    SceneDirectionType["INCREASE"] = "increase";
    SceneDirectionType["REDUCE"] = "reduce";
    SceneDirectionType["MULTIPLY"] = "multiply";
    SceneDirectionType["DIVIDE"] = "divide";
    SceneDirectionType["SET"] = "set";
    SceneDirectionType["FLAG"] = "flag";
    SceneDirectionType["UNFLAG"] = "unflag";
    SceneDirectionType["CLEAR"] = "clear";
    SceneDirectionType["RECAP"] = "recap";
    SceneDirectionType["ROLL"] = "roll dice";
    SceneDirectionType["ADD_TO_INVENTORY"] = "add inventory";
    SceneDirectionType["ADD_ITEM"] = "add item";
    SceneDirectionType["REMOVE_ITEM"] = "remove item";
    SceneDirectionType["REMOVE_FIRST"] = "remove first";
    SceneDirectionType["REMOVE_LAST"] = "remove last";
    SceneDirectionType["ADD_COLLECTION"] = "add collection";
    SceneDirectionType["CUSTOM"] = "custom";
    SceneDirectionType["REPROMPT"] = "reprompt";
    SceneDirectionType["GET_TIME"] = "get time";
    SceneDirectionType["VISUALS"] = "visuals";
    SceneDirectionType["SLOT"] = "slot";
    SceneDirectionType["BOOKMARK"] = "bookmark";
    SceneDirectionType["GO_TO"] = "go to";
    SceneDirectionType["SAVE_AND_GO"] = "save and go";
    SceneDirectionType["END"] = "ending";
    SceneDirectionType["RESTART"] = "restart";
    SceneDirectionType["RESUME"] = "resume";
    SceneDirectionType["RETURN"] = "return";
    SceneDirectionType["REPEAT"] = "repeat";
    SceneDirectionType["BACK"] = "back";
    SceneDirectionType["PAUSE"] = "pause";
    SceneDirectionType["REPEAT_REPROMPT"] = "repeat reprompt";
})(SceneDirectionType = exports.SceneDirectionType || (exports.SceneDirectionType = {}));
