// @alexa-games/sfb-util/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var BuiltInIntent;
(function (BuiltInIntent) {
    BuiltInIntent["CANCEL_INTENT"] = "AMAZON.CancelIntent";
    BuiltInIntent["HELP_INTENT"] = "AMAZON.HelpIntent";
    BuiltInIntent["STOP_INTENT"] = "AMAZON.StopIntent";
    BuiltInIntent["PAUSE_INTENT"] = "AMAZON.PauseIntent";
    BuiltInIntent["RESUME_INTENT"] = "AMAZON.ResumeIntent";
    BuiltInIntent["MORE_INTENT"] = "AMAZON.MoreIntent";
    BuiltInIntent["NAVIGATE_HOME_INTENT"] = "AMAZON.NavigateHomeIntent";
    BuiltInIntent["NAVIGATE_SETTINGS_INTENT"] = "AMAZON.NavigateSettingsIntent";
    BuiltInIntent["NEXT_INTENT"] = "AMAZON.NextIntent";
    BuiltInIntent["PAGE_UP_INTENT"] = "AMAZON.PageUpIntent";
    BuiltInIntent["PAGE_DOWN_INTENT"] = "AMAZON.PageDownIntent";
    BuiltInIntent["PREVIOUS_INTENT"] = "AMAZON.PreviousIntent";
    BuiltInIntent["SCROLL_RIGHT_INTENT"] = "AMAZON.ScrollRightIntent";
    BuiltInIntent["SCROLL_DOWN_INTENT"] = "AMAZON.ScrollDownIntent";
    BuiltInIntent["SCROLL_LEFT_INTENT"] = "AMAZON.ScrollLeftIntent";
    BuiltInIntent["SCROLL_UP_INTENT"] = "AMAZON.ScrollUpIntent";
    BuiltInIntent["REPEAT_INTENT"] = "AMAZON.RepeatIntent";
    BuiltInIntent["YES_INTENT"] = "AMAZON.YesIntent";
    BuiltInIntent["NO_INTENT"] = "AMAZON.NoIntent";
    BuiltInIntent["START_OVER_INTENT"] = "AMAZON.StartOverIntent";
    BuiltInIntent["SHUFFLE_ON_INTENT"] = "AMAZON.ShuffleOnIntent";
    BuiltInIntent["SHUFFLE_OFF_INTENT"] = "AMAZON.ShuffleOffIntent";
    BuiltInIntent["LOOP_ON_INTENT"] = "AMAZON.LoopOnIntent";
    BuiltInIntent["LOOP_OFF_INTENT"] = "AMAZON.LoopOffIntent";
    BuiltInIntent["FALLBACK_INTENT"] = "AMAZON.FallbackIntent";
})(BuiltInIntent = exports.BuiltInIntent || (exports.BuiltInIntent = {}));
;
var BuiltInSlotType;
(function (BuiltInSlotType) {
    BuiltInSlotType["NUMBER"] = "AMAZON.NUMBER";
    BuiltInSlotType["US_FIRST_NAME"] = "AMAZON.US_FIRST_NAME";
})(BuiltInSlotType = exports.BuiltInSlotType || (exports.BuiltInSlotType = {}));
var AlexaSkillsKitEvent;
(function (AlexaSkillsKitEvent) {
    AlexaSkillsKitEvent["NEW_SESSION"] = "NewSession";
    AlexaSkillsKitEvent["LAUNCH_REQUEST"] = "LaunchRequest";
    AlexaSkillsKitEvent["SESSION_ENDED_REQUEST"] = "SessionEndedRequest";
    AlexaSkillsKitEvent["UNHANDLED"] = "Unhandled";
})(AlexaSkillsKitEvent = exports.AlexaSkillsKitEvent || (exports.AlexaSkillsKitEvent = {}));
