// @alexa-games/sfb-util/dist/interaction-model-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("./constants");
const constants_2 = require("./constants");
const MAX_SLOT_SIZE = 140;
class InteractionModelBuilder {
    constructor() {
        this.invocationName = undefined;
        this.builtInIntentsBuilder = new BuiltInIntentsBuilder();
        this.intentBuilders = [];
        this.customSlotTypeBuilders = [];
        this.catchAllIntentBuilder = undefined;
    }
    withInvocationName(invocationName) {
        this.invocationName = invocationName;
        return this;
    }
    withBuiltInIntents(builtInIntentsBuilderCallback) {
        builtInIntentsBuilderCallback(this.builtInIntentsBuilder);
        return this;
    }
    withIntent(intentBuilderCallback) {
        let intentBuilder = new IntentBuilder();
        intentBuilderCallback(intentBuilder);
        this.intentBuilders.push(intentBuilder);
        return this;
    }
    withCustomSlotType(customSlotTypeBuilderCallback) {
        let customSlotTypeBuilder = new CustomSlotTypeBuilder();
        customSlotTypeBuilderCallback(customSlotTypeBuilder);
        this.customSlotTypeBuilders.push(customSlotTypeBuilder);
        return this;
    }
    withCatchAllIntent(catchAllIntentBuilderCallback) {
        this.catchAllIntentBuilder = new CatchAllIntentBuilder();
        catchAllIntentBuilderCallback(this.catchAllIntentBuilder);
        return this;
    }
    build() {
        let types = this.customSlotTypeBuilders.map(customSlotTypeBuilder => customSlotTypeBuilder.build());
        let intents = this.builtInIntentsBuilder.build();
        let customIntents = this.intentBuilders.map(intentBuilder => intentBuilder.build());
        Array.prototype.push.apply(intents, customIntents);
        if (this.catchAllIntentBuilder) {
            intents.push(this.catchAllIntentBuilder.buildIntent());
            types.push(this.catchAllIntentBuilder.buildCustomSlotType());
        }
        let interactionModel = {
            languageModel: {
                types,
                intents,
                invocationName: this.invocationName
            }
        };
        return interactionModel;
    }
}
exports.InteractionModelBuilder = InteractionModelBuilder;
class BuiltInIntentBuilder {
    constructor(intentName) {
        this.sampleUtterances = [];
        this.enabled = false;
        this.intentName = intentName;
    }
    withSampleUtterance(sampleUtterance) {
        this.sampleUtterances.push(sampleUtterance);
        return this;
    }
    withSampleUtterances(generateSampleUtterances) {
        Array.prototype.push.apply(this.sampleUtterances, generateSampleUtterances());
        return this;
    }
    enable() {
        this.enabled = true;
    }
    isEnabled() {
        return this.enabled;
    }
    build() {
        return {
            name: this.intentName,
            samples: this.sampleUtterances
        };
    }
}
class BuiltInIntentsBuilder {
    constructor() {
        this.cancelIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.CANCEL_INTENT);
        this.helpIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.HELP_INTENT);
        this.stopIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.STOP_INTENT);
        this.pauseIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.PAUSE_INTENT);
        this.resumeIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.RESUME_INTENT);
        this.moreIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.MORE_INTENT);
        this.navigateHomeIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.NAVIGATE_HOME_INTENT);
        this.navigateSettingsIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.NAVIGATE_SETTINGS_INTENT);
        this.nextIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.NEXT_INTENT);
        this.pageUpIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.PAGE_UP_INTENT);
        this.pageDownIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.PAGE_DOWN_INTENT);
        this.previousIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.PREVIOUS_INTENT);
        this.scrollRightIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.SCROLL_RIGHT_INTENT);
        this.scrollDownIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.SCROLL_DOWN_INTENT);
        this.scrollLeftIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.SCROLL_LEFT_INTENT);
        this.scrollUpIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.SCROLL_UP_INTENT);
        this.repeatIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.REPEAT_INTENT);
        this.yesIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.YES_INTENT);
        this.noIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.NO_INTENT);
        this.startOverIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.START_OVER_INTENT);
        this.shuffleOnIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.SHUFFLE_ON_INTENT);
        this.shuffleOffIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.SHUFFLE_OFF_INTENT);
        this.loopOnIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.LOOP_ON_INTENT);
        this.loopOffIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.LOOP_OFF_INTENT);
        this.fallbackIntentBuilder = new BuiltInIntentBuilder(constants_1.BuiltInIntent.FALLBACK_INTENT);
        this.builtInIntentsBuilders = [];
        this.builtInIntentsBuilders.push(this.cancelIntentBuilder);
        this.builtInIntentsBuilders.push(this.helpIntentBuilder);
        this.builtInIntentsBuilders.push(this.stopIntentBuilder);
        this.builtInIntentsBuilders.push(this.pauseIntentBuilder);
        this.builtInIntentsBuilders.push(this.resumeIntentBuilder);
        this.builtInIntentsBuilders.push(this.moreIntentBuilder);
        this.builtInIntentsBuilders.push(this.navigateHomeIntentBuilder);
        this.builtInIntentsBuilders.push(this.navigateSettingsIntentBuilder);
        this.builtInIntentsBuilders.push(this.nextIntentBuilder);
        this.builtInIntentsBuilders.push(this.pageUpIntentBuilder);
        this.builtInIntentsBuilders.push(this.pageDownIntentBuilder);
        this.builtInIntentsBuilders.push(this.previousIntentBuilder);
        this.builtInIntentsBuilders.push(this.scrollRightIntentBuilder);
        this.builtInIntentsBuilders.push(this.scrollDownIntentBuilder);
        this.builtInIntentsBuilders.push(this.scrollLeftIntentBuilder);
        this.builtInIntentsBuilders.push(this.scrollUpIntentBuilder);
        this.builtInIntentsBuilders.push(this.repeatIntentBuilder);
        this.builtInIntentsBuilders.push(this.yesIntentBuilder);
        this.builtInIntentsBuilders.push(this.noIntentBuilder);
        this.builtInIntentsBuilders.push(this.startOverIntentBuilder);
        this.builtInIntentsBuilders.push(this.shuffleOnIntentBuilder);
        this.builtInIntentsBuilders.push(this.shuffleOffIntentBuilder);
        this.builtInIntentsBuilders.push(this.loopOnIntentBuilder);
        this.builtInIntentsBuilders.push(this.loopOffIntentBuilder);
        this.builtInIntentsBuilders.push(this.fallbackIntentBuilder);
    }
    thatUsesAudioPlayerInterface() {
        this.pauseIntentBuilder.enable();
        this.resumeIntentBuilder.enable();
        return this;
    }
    thatUsesDisplayInterface() {
        this.moreIntentBuilder.enable();
        this.navigateHomeIntentBuilder.enable();
        this.navigateSettingsIntentBuilder.enable();
        this.nextIntentBuilder.enable();
        this.pageUpIntentBuilder.enable();
        this.pageDownIntentBuilder.enable();
        this.previousIntentBuilder.enable();
        this.scrollRightIntentBuilder.enable();
        this.scrollDownIntentBuilder.enable();
        this.scrollLeftIntentBuilder.enable();
        this.scrollUpIntentBuilder.enable();
        return this;
    }
    withCancelIntent(builtInIntentBuilderCallback) {
        this.cancelIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.cancelIntentBuilder);
        }
        return this;
    }
    withHelpIntent(builtInIntentBuilderCallback) {
        this.helpIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.helpIntentBuilder);
        }
        return this;
    }
    withStopIntent(builtInIntentBuilderCallback) {
        this.stopIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.stopIntentBuilder);
        }
        return this;
    }
    withPauseIntent(builtInIntentBuilderCallback) {
        this.pauseIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.pauseIntentBuilder);
        }
        return this;
    }
    withResumeIntent(builtInIntentBuilderCallback) {
        this.resumeIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.resumeIntentBuilder);
        }
        return this;
    }
    withMoreIntent(builtInIntentBuilderCallback) {
        this.moreIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.moreIntentBuilder);
        }
        return this;
    }
    withNavigateHomeIntent(builtInIntentBuilderCallback) {
        this.navigateHomeIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.navigateHomeIntentBuilder);
        }
        return this;
    }
    withNavigateSettingsIntent(builtInIntentBuilderCallback) {
        this.navigateSettingsIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.navigateSettingsIntentBuilder);
        }
        return this;
    }
    withNextIntent(builtInIntentBuilderCallback) {
        this.nextIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.nextIntentBuilder);
        }
        return this;
    }
    withPageUpIntent(builtInIntentBuilderCallback) {
        this.pageUpIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.pageUpIntentBuilder);
        }
        return this;
    }
    withPageDownIntent(builtInIntentBuilderCallback) {
        this.pageDownIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.pageDownIntentBuilder);
        }
        return this;
    }
    withPreviousIntent(builtInIntentBuilderCallback) {
        this.previousIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.previousIntentBuilder);
        }
        return this;
    }
    withScrollRightIntent(builtInIntentBuilderCallback) {
        this.scrollRightIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.scrollRightIntentBuilder);
        }
        return this;
    }
    withScrollDownIntent(builtInIntentBuilderCallback) {
        this.scrollDownIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.scrollDownIntentBuilder);
        }
        return this;
    }
    withScrollLeftIntent(builtInIntentBuilderCallback) {
        this.scrollLeftIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.scrollLeftIntentBuilder);
        }
        return this;
    }
    withScrollUpIntent(builtInIntentBuilderCallback) {
        this.scrollUpIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.scrollUpIntentBuilder);
        }
        return this;
    }
    withRepeatIntent(builtInIntentBuilderCallback) {
        this.repeatIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.repeatIntentBuilder);
        }
        return this;
    }
    withYesIntent(builtInIntentBuilderCallback) {
        this.yesIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.yesIntentBuilder);
        }
        return this;
    }
    withNoIntent(builtInIntentBuilderCallback) {
        this.noIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.noIntentBuilder);
        }
        return this;
    }
    withStartOverIntent(builtInIntentBuilderCallback) {
        this.startOverIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.startOverIntentBuilder);
        }
        return this;
    }
    withShuffleOnIntent(builtInIntentBuilderCallback) {
        this.shuffleOnIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.shuffleOnIntentBuilder);
        }
        return this;
    }
    withShuffleOffIntent(builtInIntentBuilderCallback) {
        this.shuffleOffIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.shuffleOffIntentBuilder);
        }
        return this;
    }
    withLoopOnIntent(builtInIntentBuilderCallback) {
        this.loopOnIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.loopOnIntentBuilder);
        }
        return this;
    }
    withLoopOffIntent(builtInIntentBuilderCallback) {
        this.loopOffIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.loopOffIntentBuilder);
        }
        return this;
    }
    withFallbackIntent(builtInIntentBuilderCallback) {
        this.fallbackIntentBuilder.enable();
        if (builtInIntentBuilderCallback) {
            builtInIntentBuilderCallback(this.fallbackIntentBuilder);
        }
        return this;
    }
    build() {
        let intents = [];
        this.builtInIntentsBuilders.forEach(builtInIntentsBuilder => {
            if (builtInIntentsBuilder.isEnabled()) {
                intents.push(builtInIntentsBuilder.build());
            }
        });
        return intents;
    }
}
class IntentBuilder {
    constructor() {
        this.intentName = 'CatchAllIntent';
        this.slotBuilders = [];
        this.sampleUtterances = [];
    }
    withIntentName(intentName) {
        this.intentName = intentName;
        return this;
    }
    withSlot(slotBuilderCallback) {
        let slotBuilder = new SlotBuilder();
        slotBuilderCallback(slotBuilder);
        this.slotBuilders.push(slotBuilder);
        return this;
    }
    withSampleUtterance(sampleUtterance) {
        this.sampleUtterances.push(sampleUtterance);
        return this;
    }
    withSampleUtterances(generateSampleUtterances) {
        Array.prototype.push.apply(this.sampleUtterances, generateSampleUtterances());
        return this;
    }
    build() {
        if (!this.intentName) {
            throw new Error('IntentName was not defined for intent');
        }
        let slots = this.slotBuilders.map(slotBuilder => slotBuilder.build());
        return {
            name: this.intentName,
            samples: this.sampleUtterances,
            slots
        };
    }
}
class SlotBuilder {
    constructor() {
        this.slotName = undefined;
        this.builtInSlotType = undefined;
        this.customSlotType = undefined;
    }
    withSlotName(slotName) {
        this.slotName = slotName;
        return this;
    }
    withBuiltInSlotType(builtInSlotType) {
        this.builtInSlotType = builtInSlotType;
        return this;
    }
    withCustomSlotType(customSlotType) {
        this.customSlotType = customSlotType;
        return this;
    }
    getSlotName() {
        return this.slotName;
    }
    build() {
        if (!this.slotName) {
            throw new Error('SlotName is not defined for slot');
        }
        let type;
        if (this.builtInSlotType) {
            type = this.builtInSlotType;
        }
        else if (this.customSlotType) {
            type = this.customSlotType;
        }
        else {
            throw new Error('BuiltInSlotType or CustomSlotType are not defined for slot');
        }
        return {
            name: this.slotName,
            type
        };
    }
}
class CustomSlotTypeBuilder {
    constructor() {
        this.customSlotTypeName = undefined;
        this.customValues = [];
    }
    withCustomSlotTypeName(customSlotTypeName) {
        this.customSlotTypeName = customSlotTypeName;
        return this;
    }
    withCustomValues(generateCustomValues) {
        Array.prototype.push.apply(this.customValues, generateCustomValues());
        return this;
    }
    build() {
        if (!this.customSlotTypeName) {
            throw new Error('CustomSlotTypeName is not defined for custom slot type');
        }
        if (this.customValues.length <= 0) {
            throw new Error('No custom values defined for custom slot type');
        }
        return {
            name: this.customSlotTypeName,
            values: this.buildSlotTypeValues()
        };
    }
    buildSlotTypeValues() {
        let existingSlotTypeValues = {};
        let slotTypeValues = [];
        this.customValues.forEach(value => {
            if (value) {
                let trimmedValue = value.trim().substr(0, MAX_SLOT_SIZE);
                if (trimmedValue) {
                    let trimmedLowerCaseValue = trimmedValue.toLowerCase();
                    if (!existingSlotTypeValues[trimmedLowerCaseValue]) {
                        let slotItem = this.buildSlotItem(trimmedValue);
                        slotTypeValues.push(slotItem);
                        existingSlotTypeValues[trimmedLowerCaseValue] = true;
                    }
                }
            }
        });
        return slotTypeValues;
    }
    buildSlotItem(value) {
        let slotItem = {
            id: null,
            name: {
                value: value,
                synonyms: []
            }
        };
        return slotItem;
    }
}
class CatchAllIntentBuilder {
    constructor() {
        this.catchAllIntentName = undefined;
        this.catchAllSlotBuilder = new SlotBuilder();
        this.catchAllCustomSlotTypeBuilder = new CustomSlotTypeBuilder();
        this.slotBuilders = [];
        this.sampleUtterances = [];
        this.catchAllSlotBuilder.withSlotName('CatchAll').withCustomSlotType('CatchAllSlotType');
        this.catchAllCustomSlotTypeBuilder.withCustomSlotTypeName('CatchAllSlotType');
    }
    withCatchAllIntentName(intentName) {
        this.catchAllIntentName = intentName;
        return this;
    }
    withCatchAllSlotName(slotName) {
        this.catchAllSlotBuilder.withSlotName(slotName);
        return this;
    }
    withCatchAllCustomSlotTypeName(slotTypeName) {
        this.catchAllSlotBuilder.withCustomSlotType(slotTypeName);
        this.catchAllCustomSlotTypeBuilder.withCustomSlotTypeName(slotTypeName);
        return this;
    }
    withCatchAllCustomValues(generateCustomValues) {
        this.catchAllCustomSlotTypeBuilder.withCustomValues(generateCustomValues);
        return this;
    }
    withRandomWordsAndPhrasesInCatchAllCustomValues() {
        this.catchAllCustomSlotTypeBuilder.withCustomValues(generateRandomWordsAndPhrases);
        return this;
    }
    withExceptionWordsInCatchAllCustomValues() {
        this.catchAllCustomSlotTypeBuilder.withCustomValues(generateExceptionWords);
        return this;
    }
    withExtraSlot(slotBuilderCallback) {
        let slotBuilder = new SlotBuilder();
        slotBuilderCallback(slotBuilder);
        this.slotBuilders.push(slotBuilder);
        return this;
    }
    withExtraSampleUtterance(sampleUtterance) {
        this.sampleUtterances.push(sampleUtterance);
        return this;
    }
    buildIntent() {
        if (!this.catchAllIntentName) {
            throw new Error('CatchAllIntentName was not defined for catch all intent');
        }
        let samples = ['{' + this.catchAllSlotBuilder.getSlotName() + '}'];
        Array.prototype.push.apply(samples, this.sampleUtterances);
        let slots = this.slotBuilders.map(slotBuilder => slotBuilder.build());
        slots.push(this.catchAllSlotBuilder.build());
        return {
            name: this.catchAllIntentName,
            samples,
            slots
        };
    }
    buildCustomSlotType() {
        return this.catchAllCustomSlotTypeBuilder.build();
    }
}
function generateExceptionWords() {
    let exceptionWords = [
        'Soap',
        'Sop',
        'Sap',
        'Four',
        'Pap',
        'Pop',
        'top',
        'none'
    ];
    return exceptionWords;
}
function generateRandomWordsAndPhrases() {
    // Generated by https://randomwordgenerator.com/phrase.php
    let randomPhrases = [
        'What Am I, Chopped Liver?',
        'Back to Square One',
        'Shot In the Dark',
        'What Goes Up Must Come Down',
        'Keep Your Eyes Peeled',
        'In a Pickle',
        'Hit Below The Belt',
        'Elvis Has Left The Building',
        'On the Ropes',
        'Eat My Hat',
        'Keep Your Shirt On',
        'Keep On Truckin\'',
        'My Cup of Tea',
        'Talk the Talk',
        'Mountain Out of a Molehill',
        'Fight Fire With Fire',
        'Jaws of Life',
        'Beating Around the Bush',
        'Scot-free',
        'Down And Out',
        'Every Cloud Has a Silver Lining',
        'Tough It Out',
        'Money Doesn\'t Grow On Trees',
        'Under the Weather',
        'No Ifs, Ands, or Buts',
        'Greased Lightning',
        'You Can\'t Judge a Book By Its Cover',
        'Ride Him, Cowboy!',
        'Down To The Wire',
        'Read \'Em and Weep',
        'Know the Ropes',
        'Ring Any Bells?',
        'Back To the Drawing Board',
        'Between a Rock and a Hard Place',
        'Short End of the Stick',
        'Playing For Keeps',
        'On Cloud Nine',
        'Easy As Pie',
        'Beating a Dead Horse',
        'Lickety Split',
        'Poke Fun At',
        'Like Father Like Son',
        'Two Down, One to Go',
        'Break The Ice',
        'Swinging For the Fences',
        'Cup Of Joe',
        'Drawing a Blank',
        'Hard Pill to Swallow',
        'Quality Time',
        'There\'s No I in Team',
        'The Plot Thickens',
        'Yada Yada',
        'Quick and Dirty',
        'All Greek To Me',
        'A Chip on Your Shoulder',
        'Barking Up The Wrong Tree',
        'I Smell a Rat',
        'Off One\'s Base',
        'Jumping the Gun',
        'Raining Cats and Dogs',
        'Jaws of Death',
        'Son of a Gun',
        'Down For The Count',
        'Wild Goose Chase',
        'Head Over Heels',
        'Roll With the Punches',
        'Knock Your Socks Off',
        'Down To Earth',
        'Heads Up',
        'Jack of All Trades Master of None',
        'If You Can\'t Stand the Heat, Get Out of the Kitchen',
        'Curiosity Killed The Cat',
        'No-Brainer',
        'Go Out On a Limb',
        'Jig Is Up',
        'In the Red',
        'Happy as a Clam',
        'Cut The Mustard',
        'A Piece of Cake',
        'Not the Sharpest Tool in the Shed',
        'Hear, Hear',
        'Let Her Rip',
        'Cry Over Spilt Milk',
        'Close But No Cigar',
        'Throw In the Towel',
        'Cut To The Chase',
        'Up In Arms',
        'High And Dry',
        'Love Birds',
        'Drive Me Nuts',
        'Needle In a Haystack',
        'Par For the Course',
        'A Dime a Dozen',
        'Top Drawer',
        'Birds of a Feather Flock Together',
        'It\'s Not All It\'s Cracked Up To Be',
        'Ugly Duckling',
        'Playing Possum',
        'Don\'t Count Your Chickens Before They Hatch',
        'An Arm and a Leg',
        'Hands Down',
        'Right Off the Bat',
        'Don\'t Look a Gift Horse In The Mouth',
        'A Fool and His Money are Soon Parted',
        'Give a Man a Fish',
        'Fish Out Of Water',
        'Under Your Nose',
        'Right Out of the Gate',
        'Wouldn\'t Harm a Fly',
        'On the Same Page',
        'Lovey Dovey',
        'Man of Few Words',
        'Quick On the Draw',
        'Elephant in the Room',
        'Fool\'s Gold',
        'Goody Two-Shoes',
        'Everything But The Kitchen Sink',
        'You Can\'t Teach an Old Dog New Tricks',
        'Burst Your Bubble',
        'Rain on Your Parade',
        'Wake Up Call',
        'Dropping Like Flies',
        'When the Rubber Hits the Road',
        'Flea Market',
        'Foaming At The Mouth',
        'Cry Wolf',
        'Tug of War',
        'It\'s Not Brain Surgery',
        'Long In The Tooth',
        'Mouth-watering',
        'Go For Broke',
        'Knuckle Down',
        'Fit as a Fiddle',
        'Put a Sock In It'
    ];
    // Generated by https://randomwordgenerator.com
    let randomWords = [
        'nominate',
        'name',
        'explain',
        'radio',
        'clash',
        'cool',
        'nail',
        'hallway',
        'horoscope',
        'approval',
        'feedback',
        'reverse',
        'specimen',
        'candle',
        'category',
        'cheque',
        'scratch',
        'virtue',
        'relate',
        'identity',
        'jail',
        'conflict',
        'conscience',
        'rider',
        'difficulty',
        'version',
        'far',
        'sunrise',
        'voyage',
        'drain',
        'precede',
        'widen',
        'ash',
        'example',
        'haunt',
        'cabinet',
        'warning',
        'opponent',
        'retirement',
        'relax',
        'corner',
        'neighbour',
        'error',
        'customer',
        'deprivation',
        'module',
        'afford',
        'instrument',
        'musical',
        'dead',
        'trend',
        'stream',
        'discriminate',
        'vessel',
        'moon',
        'anger',
        'quotation',
        'chew',
        'seasonal',
        'thinker',
        'quantity',
        'finger',
        'lean',
        'open',
        'registration',
        'reliable',
        'gold',
        'ambiguity',
        'project',
        'request',
        'assumption',
        'remunerate',
        'ambiguous',
        'consultation',
        'neighborhood',
        'stable',
        'coverage',
        'testify',
        'anniversary',
        'ballet',
        'drag',
        'bacon',
        'sequence',
        'perform',
        'block',
        'folk music',
        'patent',
        'definite',
        'cave',
        'begin',
        'defeat',
        'crystal',
        'damn',
        'thank',
        'decide',
        'reason',
        'telephone',
        'injury',
        'brag',
        'manufacturer',
        'digital',
        'lay',
        'period',
        'use',
        'negative',
        'leash',
        'random',
        'salad',
        'implicit',
        'compose',
        'episode',
        'split',
        'wolf',
        'acceptable',
        'cellar',
        'list',
        'estate',
        'cafe',
        'stool',
        'agenda',
        'adventure',
        'commission',
        'room',
        'girlfriend',
        'produce',
        'profession',
        'motivation',
        'home',
        'revenge',
        'easy',
        'absolute',
        'bet',
        'book',
        'contain',
        'preach',
        'green',
        'wire',
        'decrease',
        'imposter',
        'calendar',
        'jump',
        'share',
        'hemisphere',
        'deputy',
        'level',
        'capture',
        'continental',
        'owl',
        'accompany',
        'castle',
        'railroad',
        'disco',
        'monstrous',
        'spontaneous',
        'live',
        'orbit',
        'gasp',
        'science',
        'punch',
        'implication',
        'despair',
        'button',
        'location',
        'terms',
        'width',
        'fever',
        'ranch',
        'dorm',
        'performance',
        'yearn',
        'faith',
        'endure',
        'wave',
        'refund',
        'nightmare',
        'impress',
        'closed',
        'courtesy',
        'elapse',
        'legislature',
        'meadow',
        'strange',
        'finish',
        'mountain',
        'sit',
        'forest',
        'technique',
        'lip',
        'inch',
        'belong',
        'concentrate',
        'entertainment',
        'carry',
        'retiree',
        'camp',
        'scrape',
        'connection',
        'mine',
        'king'
    ];
    let randomInputs = [];
    Array.prototype.push.apply(randomInputs, randomPhrases);
    Array.prototype.push.apply(randomInputs, randomWords);
    return randomInputs;
}
let unitTests = false;
if (unitTests) {
    let interactionModelBuilder = new InteractionModelBuilder();
    let interactionModel = interactionModelBuilder
        .withInvocationName('memory wizard test')
        .withBuiltInIntents(x => x
        .withHelpIntent(x => x
        .withSampleUtterance('Help Me'))
        .thatUsesDisplayInterface()
        .withNextIntent(x => x
        .withSampleUtterance('Skip')))
        .withCustomSlotType(x => x
        .withCustomSlotTypeName('CustomSlotType')
        .withCustomValues(() => ['custom value', 'custom value 2']))
        .withIntent(x => x.withIntentName('ProceedIntent').withSampleUtterance('Proceed'))
        .withIntent(x => x.withIntentName('RepeatIntent').withSampleUtterance('Repeat'))
        .withIntent(x => x.
        withIntentName('GiveAgeIntent')
        .withSampleUtterance('My name is {Age}')
        .withSlot(x => x
        .withSlotName('Age')
        .withBuiltInSlotType(constants_2.BuiltInSlotType.NUMBER)))
        .withCatchAllIntent(x => x
        .withCatchAllIntentName('AnswerIntent')
        .withCatchAllSlotName('Answer')
        .withCatchAllCustomSlotTypeName('AnswerSlotType')
        .withCatchAllCustomValues(() => ['Yes, Okay'])
        .withRandomWordsAndPhrasesInCatchAllCustomValues()
        .withExceptionWordsInCatchAllCustomValues())
        .build();
    console.log(JSON.stringify(interactionModel));
}
