// @alexa-games/sfb-f/storyEntities/StoryAccessor.ts
import { StoryMetadata, Scene, BuiltInScenes, SceneDirection } from "../aceEntities/interfaces";

/**
 * Accessor / Helper for StoryMetadata object.
 */
export class StoryAccessor {
    protected story: {[key: string]: Scene};
    protected sourceData: StoryMetadata;

    constructor(story: StoryMetadata) {
        this.sourceData = JSON.parse(JSON.stringify(story));

        this.story = {};

        let id2Page: {[key: string]: Scene} = {};

        for (let scene of this.sourceData.scenes) {
            id2Page[scene.id.toLowerCase()] = scene;
        }

        this.story = id2Page;
    }

    /**
     * Get the scene object of the currently driving story givene the scene ID.
     * @param sceneID scene ID
     */
    public getSceneByID(sceneID: string): Scene {
        let adjustedSceneID: string = sceneID.trim().toLowerCase();

        if (this.story[adjustedSceneID] || BuiltInScenes.PauseScene.toLowerCase() == adjustedSceneID ||
            BuiltInScenes.ResumeScene.toLowerCase() == adjustedSceneID) {
            return this.story[adjustedSceneID];
        } else {
            throw `[ERROR] Cannot find the scene ${adjustedSceneID}.`; 
        }
    }

    public getSceneNarration(sceneID: string): string {
        let adjustedSceneID: string = sceneID.trim().toLowerCase();

        if (this.story[adjustedSceneID] && this.story[adjustedSceneID].contents && this.story[adjustedSceneID].contents[0]) {
            const narration = this.story[adjustedSceneID].contents[0].narration;
            if (narration) {
                return narration;
            }
        }

        throw `[ERROR] Cannot find the scene ${adjustedSceneID}.`;
    }

    public getSceneInstructions(sceneID: string): SceneDirection[] {
        let adjustedSceneID: string = sceneID.trim().toLowerCase();

        if (this.story[adjustedSceneID] && this.story[adjustedSceneID].contents && this.story[adjustedSceneID].contents[0]) {
            const instructions = this.story[adjustedSceneID].contents[0].sceneDirections;
            if (instructions) {
                return instructions;
            }
        }

        throw `[ERROR] Cannot find the scene ${adjustedSceneID}.`;
    }

    /**
     * Gets a list of Scenes within the current story.
     */
    public getAllScenes(): Scene[] {
        try {
            let storyReference: {[key: string]: Scene} = JSON.parse(JSON.stringify(this.story));
            let scenes: Scene[] = Object.keys(storyReference).map(function (value: string, index: number, array: string[]) {
                return storyReference[value];
            });

            return scenes;
        } catch (err) {
            throw err;
        }   
    }

    public async getStoryID(): Promise<string> {
        return this.sourceData.storyID;
    }

    public async getStoryMetadata(): Promise<StoryMetadata> {
        return JSON.parse(JSON.stringify(this.sourceData));
    }
}
