// @alexa-games/sfb-f/importPlugins/storyMetadataHelper.ts
import { StoryAccessor } from "../storyEntities/StoryAccessor";
import { SceneDirection, Scene, StoryMetadata } from '../aceEntities/interfaces';

/**
 * Help access and navigate the StoryMetadata used by the Driver
 */
export class StoryMetadataHelper extends StoryAccessor {
    constructor(story: StoryMetadata) {
        super(story);
    }

    /**
     * Set and replace all Scenes for the current import result.
     */
    public setAllScenes(scenes: Scene[]) {
        this.sourceData.scenes = JSON.parse(JSON.stringify(scenes));

        for (let scene of this.sourceData.scenes) {
            this.story[scene.id] = scene;
        }   
    }

    /**
     * Set and replace a scene for the current import result.
     */
    public setScene(sceneID: string, narration?: string, instructions?: SceneDirection[], customSceneProperty?: any): void {
        const adjustedSceneID: string = sceneID.trim().toLowerCase();

        const addingScene: Scene = {
            id: adjustedSceneID,
            contents: [
                {
                    narration: narration,
                    sceneDirections: instructions
                }
            ],
            customProperties: customSceneProperty
        }

        this.story[adjustedSceneID] = addingScene;
        this.sourceData.scenes.push(addingScene);
   }

   /**
    * Set and replace the narration property of the given scene.
    */
    public setSceneNarration(sceneID: string, narration: string): void {
        const scene = this.getSceneByID(sceneID);
        scene.contents[0].narration = narration;
    }

    /**
     * Set and replace instructions for the given scene.
     */
    public setSeceneInstructions(sceneID: string, instructions: SceneDirection[]): void {
        const scene = this.getSceneByID(sceneID);
        scene.contents[0].sceneDirections = instructions;
    }
    
}
