// @alexa-games/sfb-f/importPlugins/userInputHelper.ts
import { UserInput, Slot } from '../aceEntities/interfaces';
import { HandlerInput } from 'ask-sdk';

/**
 * Accessor/Helper for UserInput used as an input for the Driver's story processing.
 */
export class UserInputHelper {
    constructor(private userInput: UserInput) { }

    /**
     * Gets the HandlerInput generated by the ASK SDK for the current request
     */
    getHandlerInput(): HandlerInput | undefined {
        return this.userInput.handlerInput;
    }
    
    getInputValue(): string | undefined {
        return this.userInput.value;
    }

    setInputValue(value: string): void {
        this.userInput.value = value;
    }

    getInputSlots(): Slot[] | undefined {
        return this.userInput.slots;
    }

    setInputSlots(slots: Slot[]): void {
        this.userInput.slots = slots;
    }

    getInputIntent(): string | undefined {
        return this.userInput.intent;
    }

    setInputIntent(intent: string): void {
        this.userInput.intent = intent;
    }

    addInputSlot(slotName: string, value: string) {
        if (!this.userInput.slots) {
            this.userInput.slots = [];
        }

        this.userInput.slots.push({
            name: slotName,
            value: value
        });
    }

    clearUserInput(): void {
        delete this.userInput.intent;
        delete this.userInput.slots;
        delete this.userInput.value;
    }

    getUserInput(): UserInput {
        return this.userInput;
    }
}
