// @alexa-games/sfb-skill/src/handler/SFBRequestHandlerFactory.ts
import { DriverExtension, InstructionExtension, ImporterExtension } from '@alexa-games/sfb-f';
import { SFBRequestHandler } from './AlexaSFBRequestHandler';
import { ConfigAccessor } from '../configAccessor';

import * as path from 'path';

type ExtensionType = DriverExtension|InstructionExtension|ImporterExtension;

export class SFBRequestHandlerFactory {
    static create(event: any, context: any, extensions: ExtensionType[], configAccessor: ConfigAccessor, projectDir: string): SFBRequestHandler {
        const locale: string = event.request.locale? event.request.locale: "en-US";

        const dynamoDBTalbeName: any = configAccessor.getRequestLocalizedValue("dynamo-db-session-table-name", event);

        const storyMetadata: string = path.resolve(projectDir, "res", event.request.locale, configAccessor.getRequestLocalizedValue("abc-baked-filename", event));
                
        const pollyConfig: any = Object.assign(configAccessor.getRequestLocalizedValue("polly-config", event), {
            s3DomainName: configAccessor.getRequestLocalizedValue("s3-domain-name", event),
            bucketName: configAccessor.getRequestLocalizedValue("s3-bucket-name", event)
        });
        
        const defaultPollyConfig: any = configAccessor.getRequestLocalizedValue("default-narrator", event);
        
        const sfbHandler = new SFBRequestHandler({
                locale: locale,
                story: require(storyMetadata),
                pollyVoiceConfig: pollyConfig,
                defaultVoiceConfig: defaultPollyConfig,
                attributeTableName: dynamoDBTalbeName
            },
            extensions,
            configAccessor,
            projectDir);

        return sfbHandler;
    }
}
